ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs

```diff
@@ -93,29 +110,165 @@ data TestSetup = TestSetup
 
 instance Arbitrary TestSetup where
   arbitrary = do
-    setupD <- (/10)         <$> choose   (1, 10)
-    setupK <- SecurityParam <$> choose   (2, 5)
+    setupD <- (/10)         <$> choose (1, 10)
+    setupK <- SecurityParam <$> choose (2, 6)
 
     setupSlotLengthByron   <- arbitrary
     setupSlotLengthShelley <- arbitrary
 
-    setupTestConfig <- arbitrary
+    setupTestConfig <- do
+      tc <- arbitrary
+
+      -- This test has more slots than most, so we de-emphasize the relatively
+      -- expensive n=5 case. For example:
+      --
+      -- > 250 tests with 30% 2, 30% 3, 30% 4, and 10% 5 took 500s
+      -- > 250 tests with 100% 5 took 1000s
+      ncn <- frequency [(9, choose (2, 4)), (1, pure 5)]
+
+      let NumSlots t = numSlots tc
+      pure tc
+        { numCoreNodes = NumCoreNodes ncn
+        , numSlots     =
+            NumSlots $
+            min 150 $
+            -- make it very likely that there are multiple epochs
+            numByronEpochs * 10 * maxRollbacks setupK + div t 2
+        }
+
+    let TestConfig{numCoreNodes, numSlots} = setupTestConfig
 
     setupByronLowerBound <- arbitrary
     setupHardFork        <- frequency [(9, pure True), (1, pure False)]
+    setupPartition       <- genPartition numCoreNodes numSlots setupK
 
     pure TestSetup
       { setupByronLowerBound
       , setupD
       , setupHardFork
       , setupK
+      , setupPartition
       , setupSlotLengthByron
       , setupSlotLengthShelley
-      , setupTestConfig
+      , setupTestConfig = setupTestConfig
+        { nodeTopology =
+            -- use a mesh topology so the partition classes definitely remain
+            -- internally connected
+            meshNodeTopology numCoreNodes
+        }
       }
 
   -- TODO shrink
 
+-- | Generate 'setupPartition'
+genPartition :: NumCoreNodes -> NumSlots -> SecurityParam -> Gen Partition
+genPartition (NumCoreNodes n) (NumSlots t) (SecurityParam k) = do
+    let ultimateSlot :: Word64
+        ultimateSlot = assert (t > 0) $ t - 1
+
+        crop :: Word64 -> Word64
+        crop s = min ultimateSlot s
+
+    -- Fundamental plan: all @MsgRollForward@ sent during the partition only
+    -- arrive at the onset of slot after it. The partition begins at the onset
+    -- of @firstSlotIn@ and ends at the onset of @firstSlotAfter@.
+
+    -- FACT (A) The leader of @firstSlotAfter@ will forge before fully reacting
+    -- to the @MsgRollForward@s that just arrived, due to a race in the test
+    -- infrastructure that is consistently won by the forge. Thus the two
+    -- class's unique extensions consist of the blocks forged in the slots from
+    -- @firstSlotIn@ to @firstSlotAfter@ /inclusive/.
+
+    -- how long the partition should last (this may be 'crop'ped if it's near
+    -- the end of the test)
```

Comment:
Stray comment?

Reply:
It's just the summary line at the start of a large comment. I've made that more obvious.